cmd/snap/cmd_userd.go

```diff
@@ -60,12 +62,43 @@ func init() {
 	cmd.hidden = true
 }
 
+func maybeFixupUsrSnapPermissions() error {
+	usr, err := userCurrent()
+	if err != nil {
+		return err
+	}
+
+	usrSnapDir := filepath.Join(usr.HomeDir, dirs.UserHomeSnapDir)
+
+	// restrict the user's "snap dir", i.e. /home/$USER/snap, to be private with
+	// permissions o0700 so that other users cannot read the data there, some
+	// snaps such as chromium etc may store secrets inside this directory
+	// note that this operation is safe since `userd --autostart` runs as the
+	// user so there is no issue with this modification being performed as root,
+	// and being vulnerable to symlink switching attacks, etc.
+	if err := os.Chmod(usrSnapDir, 0700); err != nil {
+		// if the dir doesn't exist for some reason (i.e. maybe this user has
+		// never used snaps but snapd is still installed) then ignore the error
+		if !os.IsNotExist(err) {
+			return fmt.Errorf("failed to restrict user snap home dir: %v", err)
+		}
+	}
+
+	return nil
+}
+
 func (x *cmdUserd) Execute(args []string) error {
 	if len(args) > 0 {
 		return ErrExtraArgs
 	}
 
 	if x.Autostart {
+		// autostart is called when starting the graphical session, use that as
+		// an opportunity to fix ~/snap permission bits
+		if err := maybeFixupUsrSnapPermissions(); err != nil {
+			fmt.Fprintf(os.Stderr, "failed to fixup ~/snap permissions: %v\n", err)
```

Comment:
It could be useful to test this path, to make sure that the errors from fixup are ignored/logged even if we refactor the userd code in the future.

Reply:
this is done now